### ISSUE TITLE
Missing SH7059 CSV's in ghidra_helpers

### ISSUE THREAD

**Issue by jumiduss** [fenugrec/nissutils#20]
It seems like there is a linked file that doesn't exist in:
https://github.com/fenugrec/nissutils/blob/755e449fe862eb7ec22bb48fa1453bae2986b9eb/ghidra_helpers/nissan_load.py#L30

After searching through the repo, I haven't been able to find the file to link these to. Figured it was possible that it's missing, or the note: "typo in DS", could have been meaningful. If extra work is needed I may be able to test / fix the problems to push these tables out. 

Cheers

PS: Thanks for the thousands of hours of work on these things. I've been able to pickup skills by following your posts and forums discussions.

**Comment by fenugrec** [comment]
> or the note: "typo in DS", could have been meaningful.

That referred to the rom/ram address ranges, IIRC they were incorrect somewhere in the DS.

As for the *7059.csv files, I clearly didn't test the script on 7059 ROMs. Someone should go over the DS to make sure the periph regs and interrupts are either "identical enough" to the 7058 or if they need their own file. 7058 and 7059 share the same datasheet so I think any difference would be inconsequential for the purposes of this script. If you want to take a look first that would be great, otherwise I'll fix that line so it points to the *7058.csv files.

**Comment by jumiduss** [comment]
I'm currently looking into the register/interrupt info mentioned.

Also, I'm doing a small write-up for some additional information that may be useful for further development on the SH7059 processor if it ever becomes needed for development.

Thanks for the quick response.

(Apologies if this response is misplaced, I'm not well-read GitHub-issue-tracking etiquette)

**Comment by jumiduss** [comment]
Update #1:
Registers line up accordingly. There were updates to a select quantity of bit names for some registers, but that seems irrelevant in this application. 

The other thing to note is the hardware documentation states that the SSU Registers were added (as a result of SH7059, or both SH7058S and SH7059), but you have it included already in regs_7058.csv , which is possibly the error mentioned earlier. Additional registers in 7058 are included that may not exist on the hardware when they combined both on the datasheet.

There are definitely additional Interrupt Vectors due to the SSU, but I except it's possible something similar happened in your IVT for 7058. Checking this now.

TLDR: regs_7058.csv seems okay to link directly to SH7059 due to the additional registers being lumped into SH7058 on DS.

**Comment by jumiduss** [comment]
Update 2:
Slightly unsure how to check IVT.
In your ivt_7055.csv, you include a reset and manual reset. These are included as exception vectors and not interrupts within the SH7059 documentation. I saw two ways to proceed: 
1. Include only explicitly referenced interrupts.
2. Include all exceptions. 

Attached are the files with \*modifications\* and (removals) noted for each path stated. Also included are the finalized changes made without the marks. I would assume listing all exceptions is preferable as in files 2.*. Modifications for path 1) also apply to 2)
Changes: for SH7059:
1 Changes: Interrupt Only
- MTAD0_ADT0 (Multi Trigger A/D 0)
- MTAD1_ADT1 (Multi Trigger A/D 1)
[1.Non_Noted.ivt_7059.csv](https://github.com/user-attachments/files/18222834/1.Non_Noted.ivt_7059.csv)
[1.Noted.ivt_7059.csv](https://github.com/user-attachments/files/18222835/1.Noted.ivt_7059.csv)
2 Changes: Added or Missing Exceptions in SH7059
- 20 Exceptions reserved by the system
- Splitting the Power on reset and Manual reset based on source (PC or SP)
[2.Non_Noted.ivt_7059.csv](https://github.com/user-attachments/files/18222836/2.Non_Noted.ivt_7059.csv)
[2.Noted.ivt_7059.csv](https://github.com/user-attachments/files/18222837/2.Noted.ivt_7059.csv)

Note: SSU Interrupts seem to piggyback off SCI0 and SCI2, so no changes may be necessary unless you want a description change to note possible source of the exception.

Hope this helps!  (Additional SH7059 In next comment)

**Comment by jumiduss** [comment]
I would assume these are notable differences that (may be) significant for those going further into disassembling SH7059 (compared to the relatively similar specs on the SH7058 and SH7058S):
- ROM | 1.5 MB Flash Memory (Compared to 1 MB) | Included in an addition set of {4 x 256 KB Blocks} and a reduction of {4 x128 Kb blocks}
- RAM increased to 80 Kb (from 48 Kb)
- RAM Areas used by boot mode are:
    - 3 Kbytes starting at H'FFFF0000,
    - 4 Kbytes starting at H'FFFFB000,
    - 128 Kbytes from H'FFFFBF80 -> H'FFFFBFFF (used as the stack)
- Increased User boot MAT: 12 Kbytes (from 8 Kbytes)
    - Address H'00,0000 to H'00,2FFF
- Initial Values of Status Registers (SH7059)
    - SDSR: H'0F01
    - SDIDR(Fixed Code): H'0800B447
- Flash Memory Emulation in RAM Extended for EB0 -> EB15 (extending by one leading hex digit from SH7058)
    - H'000000
    - H'001000
    - ...
    - H'17FFFF
- Increased NOP Instructions needed after SCO to 8 (from 4) after setting bit to 1
- Registers are now initialized in BOTH "hardware standby mode" AND "software standby mode"
- Doubled the clock frequency from 40 MHz -> 80 Mhz (12.5 ns execution speed)
- Bi-Directional XTAL <-> Clock pulse Generator
- Software Standby Mode returns to "Power-On reset state" after "NMI Pin is Set: 0 -> 1" | While SH7058 returns to "Execution processing state" without pin trigger.
- When accessing the VBR, an address error now occurs when accessing the vector table (previously the stack) during exception processing.

Thanks again!

**Comment by fenugrec** [comment]
> There were updates to a select quantity of bit names for some registers, but that seems irrelevant in this application.

Agreed.

> Additional registers in 7058 are included that may not exist on the hardware when they combined both on the datasheet.

That's fine, someone working very closely at this level of detail will have the datasheet handy anyway.


> you include a reset and manual reset. These are included as exception vectors and not interrupts

Remember, the intent of this script regarding vectors, is simply to name them mostly correctly, and to mark them as pointers to code, so that ghidra's analysis has the best possible coverage. So, while the the 705x datasheet makes a distinction, for code analysis it doesn't really matter : both are a location in ROM that hold a pointer to code.

You'll note that I have the Power-on reset defined at the end of the csv; that is on purpose so the name overrides previous names which happens when i.e. unused/reserved vectors all point to the same POR code.

I looked at your files, thanks for going over all that.

I did debate whether or not to include the Reserved exceptions, but in my experience there is never anything interesting in there so I left them out.

> MTAD0_ADT0 (Multi Trigger A/D 0)
MTAD1_ADT1 (Multi Trigger A/D 1)

Good catch, those were indeed missing.


> I would assume these are notable differences that (may be) significant [...]

Well, some of those are already handled by the script (ROM / RAM size), other details I consider out of scope or just not useful inside ghidra.

All right, so I think I'll keep the registers combined in the 7058.csv file and the IVT in 7055.csv for ease of maintenance (the 7058/7059-only MTAD are labelled as such).

> some questions about your career/job path that got you into this

Heh well, I'm afraid there's not much to say - it's a longtime hobby that took unreasonable proportions for a while; my background and work is in a completely unrelated field.
